File: src/main/java/cspom/constraint/GeneralConstraint.java
package cspom.constraint;

import java.util.Arrays;
import java.util.List;

import javax.script.ScriptException;

import com.google.common.base.Joiner;

import cspom.Evaluator;
import cspom.variable.CSPOMVariable;

public final class GeneralConstraint implements CSPOMConstraint {

    public GeneralConstraint(final String name, final String description,
            final String parameters, final CSPOMVariable... scope) {
        super(name, description, parameters, scope);
    }

    public GeneralConstraint(final String description, final String parameters,
            final CSPOMVariable... scope) {
        super(description, parameters, scope);
    }

    public GeneralConstraint(final String description, final String parameters,
            final List<CSPOMVariable> scope) {
        super(description, parameters, scope);
    }

    public String toString() {
        final StringBuilder stb = new StringBuilder();
        stb.append(getDescription());
        if (getParameters() != null) {
            stb.append('{').append(getParameters()).append('}');
        }

        int iMax = getScope().size() - 1;
        stb.append('(');
        for (int i = 0;; i++) {
            stb.append(getVariable(i));
            if (i == iMax) {
                return stb.append(')').toString();
            }
            stb.append(", ");
        }
    }

    @Override
    public boolean evaluate(final Object[] tuple) {
        final StringBuilder stb = new StringBuilder();
        if (getParameters() != null) {
            stb.append("p_");
        }
        stb.append(getDescription()).append('(');
        Joiner.on(", ").appendTo(stb, Arrays.asList(tuple));

        if (getParameters() != null) {
            stb.append(", ").append(getParameters());
        }

        try {
            return Evaluator.evaluate(stb.append(')').toString());
        } catch (ScriptException e) {
            throw new IllegalStateException(e);
        }

    }
}

File: src/main/scala/cspom/constraint/GeneralConstraint.scala
package cspom.constraint

import javax.script.ScriptException
import cspom.variable.CSPOMVariable
import cspom.Evaluator

class GeneralConstraint(
  val predicate: Predicate,
  scope: Seq[CSPOMVariable])
  extends CSPOMConstraint(predicate.function, scope) {

  require(scope.nonEmpty, "The constraint must involve at least one variable")
  //  def this(description: String, parameters: String, scope: CSPOMVariable[_]*) =
  //    this(description = description, parameters = parameters,
  //      scope = scope.toList)

  def this(func: String, params: Any, scope: CSPOMVariable*) =
    this(Predicate(func, Some(params)), scope)

  def this(func: String, scope: CSPOMVariable*) =
    this(Predicate(func, None), scope)

  override def toString = {
    val stb = new StringBuilder
    stb append description
    stb.append(predicate.optParameters)
    scope.addString(stb, "(", ", ", ")").toString
  }

  override def evaluate(tuple: Seq[_]): Boolean = {
    val stb = new StringBuilder();
    if (predicate.parameters.isDefined) {
      stb append "p_"
    }
    stb append description append '('

    tuple.addString(stb, ", ")

    predicate.parameters foreach {
      case s: Seq[Any] =>
        stb.append(", ")
        s.addString(stb, ", ")
      case p: Any =>
        stb.append(", ").append(p)
    }

    Evaluator.evaluate((stb append ')').toString);
  }

  override def replacedVar(which: CSPOMVariable, by: CSPOMVariable) = {
    new GeneralConstraint(predicate,
      scope map { v => if (v == which) by else v })
  }

}